File: pulsar-common/src/main/java/ai/platon/pulsar/common/proxy/ProxyUpdateThread.java
package ai.platon.pulsar.common.proxy;

import ai.platon.pulsar.common.NetUtil;
import ai.platon.pulsar.common.config.CapabilityTypes;
import ai.platon.pulsar.common.config.ImmutableConfig;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.concurrent.TimeUnit;

public class ProxyUpdateThread extends Thread {

    protected static final Logger LOG = LoggerFactory.getLogger(ProxyUpdateThread.class);
    private final ImmutableConfig conf;
    protected long updatePeriod;
    private ProxyPool proxyPool;

    public ProxyUpdateThread(ProxyPool proxyPool, ImmutableConfig conf) {
        this.conf = conf;
        this.proxyPool = proxyPool;
        this.updatePeriod = conf.getInt("http.proxy.pool.update.period", 10 * 1000);
        this.setDaemon(true);
    }

    @Override
    public void run() {
        if (proxyPool == null) {
            LOG.error("proxy manager must not be null");
            return;
        }

        try {
            int tick = 0;
            while (true) {
                if (tick % 20 == 0) {
                    LOG.debug("updating proxy pool...");
                }

                if (tick % 20 == 0) {
                    updateProxyFromMaster();
                }

                long start = System.currentTimeMillis();
                proxyPool.recover(100);
                long elapsed = System.currentTimeMillis() - start;

                // too often, enlarge review period
                if (elapsed > updatePeriod) {
                    LOG.info("it costs {} millis to check all retired proxy servers, enlarge the check interval", elapsed);
                    updatePeriod = elapsed * 10;
                }

                proxyPool.reloadIfModified();
                Thread.sleep(updatePeriod);

                ++tick;
            }
        } catch (InterruptedException e) {
            LOG.error(e.toString());
        }
    }

    private void updateProxyFromMaster() {
        String host = conf.get(CapabilityTypes.PULSAR_MASTER_HOST, "localhost");
        int port = conf.getInt(CapabilityTypes.PULSAR_MASTER_PORT, 8081);
        String url = "http://" + host + ":" + port + "/proxy/download";

        try {
            String hostname = NetUtil.getHostname();

            // Update only if this is not the master
            if (!host.equals(hostname)) {
                String file = "synced.proxies.txt";
                Path target = Paths.get(proxyPool.getArchiveDir().toString(), file);
                String cmd = "wget " + url + " -O " + target.toString() + " > /dev/null 2>&1";
                Process p = Runtime.getRuntime().exec(cmd);

                Path link = Paths.get(proxyPool.getEnabledDir().toString(), file);
                Files.createSymbolicLink(link, target);
            }
        } catch (IOException e) {
            LOG.error(e.toString());
        }
    }

    public static void main(String[] args) {
        ImmutableConfig conf = new ImmutableConfig();

        ProxyPool proxyPool = new ProxyPool(conf);
        ProxyUpdateThread updateThread = new ProxyUpdateThread(proxyPool, conf);
        updateThread.start();

        while (!proxyPool.isEmpty()) {
            ProxyEntry proxy = proxyPool.poll();

            if (proxy.testNetwork()) {
                LOG.debug("proxy : {} is available", proxy);
            } else {
                LOG.debug("proxy : {} is not available", proxy);
            }

            try {
                TimeUnit.SECONDS.sleep(3);
            } catch (InterruptedException ignored) {
            }
        }
    }
}

File: pulsar-common/src/main/kotlin/ai/platon/pulsar/common/proxy/ProxyUpdateThread.kt
package ai.platon.pulsar.common.proxy

import ai.platon.pulsar.common.NetUtil
import ai.platon.pulsar.common.config.CapabilityTypes
import ai.platon.pulsar.common.config.ImmutableConfig
import org.apache.commons.io.FileUtils
import org.slf4j.LoggerFactory
import java.io.IOException
import java.net.URL
import java.nio.file.Files
import java.nio.file.Paths
import java.time.Duration
import java.time.Instant
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

class ProxyUpdateThread(private val proxyPool: ProxyPool, private val conf: ImmutableConfig) : Thread(), AutoCloseable {
    private var updatePeriod = conf.getDuration("http.proxy.pool.update.period", Duration.ofSeconds(20))
    private val isClosed = AtomicBoolean()

    override fun close() {
        isClosed.set(true)
    }

    init {
        this.isDaemon = true
    }

    override fun run() {
        try {
            var tick = 0
            while (!isClosed.get()) {
                if (tick % 20 == 0) {
                    log.debug("Updating proxy pool...")
                }

                if (tick % 20 == 0) {
                    tryUpdateProxyFromMaster()
                }

                val start = Instant.now()
                proxyPool.recover(100)
                val elapsed = Duration.between(start, Instant.now())

                // too often, enlarge review period
                if (elapsed > updatePeriod) {
                    log.info("It costs {} to check all retired proxy servers, enlarge the check interval", elapsed)
                    updatePeriod.plus(elapsed)
                }

                proxyPool.reloadIfModified()

                TimeUnit.MILLISECONDS.sleep(updatePeriod.toMillis())

                ++tick
            }
        } catch (e: InterruptedException) {
            log.error(e.toString())
        }
    }

    private fun tryUpdateProxyFromMaster() {
        val host = conf.get(CapabilityTypes.PULSAR_MASTER_HOST, "localhost")
        val port = conf.getInt(CapabilityTypes.PULSAR_MASTER_PORT, 8081)

        val thisHost = NetUtil.getHostname()
        if (host == thisHost) {
            return
        }

        val url = "http://$host:$port/proxy/download"
        if (!NetUtil.testNetwork(host, port)) {
            return
        }

        try {
            val filename = "synced.proxies.txt"
            val target = Paths.get(proxyPool.archiveDir.toString(), filename)
            val symbolLink = Paths.get(proxyPool.enabledDir.toString(), filename)

            Files.deleteIfExists(target)
            Files.deleteIfExists(symbolLink)

            FileUtils.copyURLToFile(URL(url), target.toFile())
            Files.createSymbolicLink(symbolLink, target)
        } catch (e: IOException) {
            log.error(e.toString())
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(ProxyUpdateThread::class.java)
    }
}

fun main() {
    val log = LoggerFactory.getLogger(ProxyUpdateThread::class.java)

    val conf = ImmutableConfig()

    val proxyPool = ProxyPool(conf)

    val updateThread = ProxyUpdateThread(proxyPool, conf)
    updateThread.start()

    while (true) {
        val proxy = proxyPool.poll()

        if (proxy != null) {
            if (proxy.testNetwork()) {
                log.info("Proxy is available | {} ", proxy)
            } else {
                log.info("Proxy is unavailable | {}", proxy)
            }

            proxyPool.offer(proxy)
        }

        TimeUnit.SECONDS.sleep(5)
    }
}
